src/geojson.rs

```diff
@@ -0,0 +1,189 @@
+use geojson::{self, GeoJson, Geometry, JsonObject, JsonValue};
+use osmpbfreader::{OsmId, OsmObj, Ref, Way};
+
+use crate::{filter, util::FloatTuple};
+
+pub fn write(objs: &BTreeMap<OsmId, OsmObj>, out: impl io::Write) -> Result<()> {
+    // Use a buffered writer to amortize flushes.
+    let mut buffer = BufWriter::new(out);
+
+    for relation in objs.values().filter(|obj| filter::by_target(obj)) {
+        if let Some(feature) = to_feature(relation, objs) {
+            let serialized = feature.to_string();
+            writeln!(buffer, "{serialized}")?;
+        }
+    }
+
+    Ok(())
+}
+
+fn to_feature(obj: &OsmObj, all_objs: &BTreeMap<OsmId, OsmObj>) -> Option<geojson::GeoJson> {
+    let tags = obj.tags();
+    let name = tags.get("name")?;
+    let admin_level = tags.get("admin_level")?;
+    let ars = tags.get("de:regionalschluessel")?;
+
+    let mut properties = JsonObject::new();
+    properties.insert(
+        String::from("adminLevel"),
+        JsonValue::from(admin_level.parse::<u8>().ok()?),
+    );
+    properties.insert(String::from("ars"), JsonValue::from(ars.as_str()));
+    properties.insert(String::from("name"), JsonValue::from(name.as_str()));
+    properties.insert(
+        String::from("osm:id"),
+        JsonValue::from(obj.relation()?.id.0),
+    );
+    properties.insert(String::from("osm:type"), JsonValue::from("relation"));
+
+    Some(GeoJson::Feature(geojson::Feature {
+        geometry: Some(Geometry::new(as_polygon(obj, all_objs)?)),
+        id: Some(geojson::feature::Id::Number(
+            serde_json::value::Number::from(obj.relation()?.id.0),
+        )),
+        properties: Some(properties),
+        ..geojson::Feature::default()
+    }))
+}
+
+fn as_polygon(obj: &OsmObj, all_objs: &BTreeMap<OsmId, OsmObj>) -> Option<geojson::Value> {
+    let to_coords = |way: &Way| -> Option<Vec<geojson::Position>> {
+        way.nodes
+            .iter()
+            .map(|node_id| {
+                let node = all_objs.get(&OsmId::Node(*node_id))?;
+                Some(vec![
+                    f64::from(node.node()?.decimicro_lon) / 10_000_000.0,
+                    f64::from(node.node()?.decimicro_lat) / 10_000_000.0,
+                ])
+            })
+            .collect()
+    };
+
+    let linestrings: Vec<Vec<geojson::Position>> = obj
+        .relation()?
+        .refs
+        .iter()
+        .filter_map(|child: &Ref| {
+            // todo treat 'inner' and contained relations as well
+            if matches!(child.role.as_str(), "outer") {
+                Some(to_coords(all_objs.get(&child.member)?.way()?)?)
+            } else {
+                None
+            }
+        })
+        .collect();
+
+    // todo report missing geometry or broken linering
+    let mut linering = create_continuous_linering(&linestrings).ok()?;
+
+    // respect right hand rule
+    if is_clockwise(&linering) {
+        linering.reverse();
+    }
+
+    Some(geojson::Value::Polygon(vec![linering]))
```

Comment:
Nit: Ich würde hier über das `Result` mappen, vielleicht nur um das `mut` zu verhindern, z.B.

```suggestion
    // todo report missing geometry or broken linering

    create_continuous_linering(&linestrings)
        .map(|mut l| {
            // respect right hand rule
            if is_clockwise(&l) {
                l.reverse();
                l
            } else {
                l
            }
        })
        .map(|l| geojson::Value::Polygon(vec![l]))
        .ok()
```

Bin mir aber auch nicht sicher, dass das leserlicher ist.

Reply:
hä? aber du hast doch das mut auch bei deiner Lösung?

Reply:
In kleinerem scope. Das ist ein pattern (finde nur den namen gerade nicht), dass man in einem kleineren scope etwas `mut` macht, aber es dann draussen nicht mehr `mut` ist. Ein bisschen egal hier, resolving.